arroyo/backends/abstract.py

```diff
@@ -175,11 +187,69 @@ def member_id(self) -> str:
         raise NotImplementedError
 
 
+class ProducerFuture(Protocol, Generic[T]):
+    """
+    An abstract interface for a kind of Future. Stdlib futures are too slow to
+    construct, so we use these.
+    """
+
+    def done(self) -> bool:
+        ...
+
+    def result(self, timeout: float | None = None) -> T:
+        """
+        Return result or raise exception. May block, but does not have to.
+        """
+        ...
+
+    def set_result(self, result: T) -> None:
+        ...
+
+    def set_exception(self, exception: Exception) -> None:
+        ...
+
+
+class SimpleProducerFuture(Generic[T]):
+    """
+    A stub for concurrent.futures.Future that does not construct any Condition
+    variables, therefore is faster to construct.
+    """
+
+    def __init__(self) -> None:
+        self.result_value: T | None = None
+        self.result_exception: Exception | None = None
+
+    def done(self) -> bool:
+        return self.result_value is not None
+
+    def result(self, timeout: float | None = None) -> T:
+        if self.result_exception is not None:
+            raise self.result_exception
+
+        if timeout is not None:
+            deadline = time.time() + timeout
+        else:
+            deadline = None
+
+        while deadline is None or time.time() < deadline:
+            if self.result_value is not None:
+                return self.result_value
+            time.sleep(0.1)
```

Comment:
Do we want to make this interval configurable?

Reply:
added a comment, but basically I only wrote this to get the tests to pass. if this becomes performance sensitive you are supposed to use the stdlib future (and that's why it's the default)

SimpleProduceFuture is only good if you use `result(timeout=0)` or call `result()` when you know the future is `done()` (as we do in produce step)

Reply:
I have some concerns with this implementation. We are providing an implementation of the Future we ask not to use in prod.
What about not using a Future implementation at all? 
If there was no max buffer size in the producer we could jsut call the `produce` method as much as we wanted. OBviously we cannot do that. We need to wait on the callbacks. Why not just keeping a counter with the number of pending callbacks that gets incremented each time a new message is produced and decremented each time a callback is received. 
When the counter reaches the same size as the buffer we start rejecting messages. 
It seems simpler, safer and does not have this concern.

Reply:
then this is a larger refactor where we would have to support two APIs or break the futures-based one. A counter would be nice if the produce step used the confluent producer directly.